content/number-theory/Eratosthenes.h

```diff
@@ -0,0 +1,36 @@
+/**
+ * Author: Jakob Kogler, chilli, pajenegod
+ * Date: 2020-04-12
+ * License: CC0
+ * Description: Prime sieve for generating all primes up to LIM.
+ * Status: Stress-tested
+ * Time: LIM=1e9 $\approx$ 1.5s
+ * Details: Despite its n log log n complexity, segmented sieve is still faster
+ * than other options, including bitset sieves and linear sieves. This is
+ * primarily due to its low memory usage, which reduces cache misses. This
+ * implementation skips even numbers.
+ *
+ * Benchmark can be found here: https://ideone.com/e7TbX4
+ */
+#pragma once
+
+const int LIM = 1e6;
+bitset<LIM> isPrime;
+vi eratosthenes() {
+	const int S = round(sqrt(LIM)), R = LIM / 2;
+	vi pr({2}), sieve(S + 1); pr.reserve(LIM / (int)log(LIM));
+	vector<array<int, 2>> cp;
+	for (int i = 3; i <= S; i += 2) if (!sieve[i]) {
+		cp.push_back({i, i * i / 2});
+		for (int j = i * i; j <= S; j += 2 * i) sieve[j] = 1;
+	}
+	for (int L = 1; L <= R; L += S) {
+		array<bool, S> block{};
+		for (auto &[p, idx] : cp)
+			for (int i=idx; i < S+L; idx = (i+=p)) block[i-L] = 1;
```

Comment:
This line is much shorter as `for (; idx < S + L; idx += p) block[idx-L] = 1;`?
Be careful. I haven't tested it.

Reply:
Unfortunately, this was done on purpose :'(  Since `idx` is a reference into the `cp` array, accessing it is slower than accessing a local variable.

Ideally the compiler would be able to hoist it out of the loop, but it seems that the mixed load/store instructions confuse the compiler. The way it's written, there's only a single load to `idx` and then repeated stores, which the compiler seems to optimize fine.

It causes a `~200ms` difference in our testing - I'll write a note about it.